File: cpp/train.cpp
#include <iostream>
#include <string>

#include "linear_algebra.h"
#include "neural_network.h"

int main() {
//////////////////////////////////////////////////////////////////
// Load Data
    std::cout << "Loading Data..." << std::endl;
    // // Define matrix dimensions
    // int X_train_rows = 60000;
    // int X_train_cols = 784;
    // int Y_train_rows = 60000;
    // int Y_train_cols = 10;
    // 
    // // Initialize matrices
    // Matrix X_train(X_train_rows, X_train_cols);
    // Matrix Y_train(Y_train_rows, Y_train_cols);
    // 
    // // Read data
    // read_csv("data/X_train.csv", &X_train);
    // read_csv("data/Y_train.csv", &Y_train);

    // Define matrix dimensions
    int num_rows = 1000;
    int X_train_rows = num_rows;
    int X_train_cols = 784;
    int Y_train_rows = num_rows;
    int Y_train_cols = 10;

    // Initialize matrices
    Matrix X_train(X_train_rows, X_train_cols);
    Matrix Y_train(Y_train_rows, Y_train_cols);

    // Read data
    read_csv_limited("data/X_train.csv", &X_train,
                     0, X_train_rows, X_train_rows, 784);
    read_csv_limited("data/Y_train.csv", &Y_train,
                     0, Y_train_rows, Y_train_rows, 10);
    
    // Print Shape
    std::cout << "X_train: (" 
              << X_train.rows << ", " 
              << X_train.cols << ")" 
              << std::endl;

    std::cout << "Y_train: (" 
              << Y_train.rows << ", " 
              << Y_train.cols << ")" 
              << std::endl;

//////////////////////////////////////////////////////////////////
// Transpose Data
    Matrix* X_train_T = transpose_matrix(&X_train);
    Matrix* Y_train_T = transpose_matrix(&Y_train);

//////////////////////////////////////////////////////////////////
// Normalize X values
    normalize_matrix(X_train_T, 0, 255);

//////////////////////////////////////////////////////////////////
// Initialize Neural Network
    NeuralNetwork nn;
    nn.add_layer(new Layer(784, 200, "Sigmoid"));
    nn.add_layer(new Layer(200, 200, "Sigmoid"));
    nn.add_layer(new Layer(200, 10, "Softmax"));

    std::cout << "Training..." << std::endl;
    // int epochs = 1000;
    int epochs = 100;
    double learning_rate = 0.1;
    std::string loss = "CatCrossEntropy";
    std::string history_path = "models/cpp_history.csv";
    nn.train(*X_train_T,
             *Y_train_T,
             epochs,
             learning_rate,
             loss,
             history_path);

    std::cout << "Saving Model..." << std::endl;
    nn.save_config("models/cpp_config.csv");
    nn.save_weights("models/cpp_weights.csv");

    return 0;
}

File: python/train.py
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

from neural_network import Layer, NeuralNetwork
from neural_network import Sigmoid, Softmax, CatCrossEntropy

####################################################################
# Load Data
print("Loading Data...")
X_train = pd.read_csv('data/X_train.csv', header=None).to_numpy()
Y_train = pd.read_csv('data/Y_train.csv', header=None).to_numpy()
# X_train = pd.read_csv('data/X_train.csv', header=None).head(10000).to_numpy()
# Y_train = pd.read_csv('data/Y_train.csv', header=None).head(10000).to_numpy()
# 
print(f"X_train: {X_train.shape}")
print(f"Y_train: {Y_train.shape}")

####################################################################
# Transpose Data
X_train = X_train.T
Y_train = Y_train.T

####################################################################
# Normalize X values
X_train = X_train / 255.

####################################################################
nn = NeuralNetwork()

batch_size = 8
nn.add_layer(Layer(784, 200, Sigmoid(), batch_size))
nn.add_layer(Layer(200, 200, Sigmoid(), batch_size))
nn.add_layer(Layer(200, 10, Softmax(), batch_size))

print("Training...")
nn.train(X_train,
         Y_train,
         # epochs=200,
         epochs=3,
         learning_rate=0.1,
         loss=CatCrossEntropy(),
         optimizer="mini_batch_gradient_descent",
         batch_size=batch_size,
         history_path="models/python_history.csv")

print("Saving Model...")
nn.save_config("models/python_config.csv")
nn.save_weights("models/python_weights.csv")
